packages/shared/lib/services/getting-started.service.ts

```diff
@@ -0,0 +1,288 @@
+
+        return updated ? Ok(updated) : Err('failed_to_update_getting_started_progress');
+    } catch (err) {
+        return Err(new Error('failed_to_update_getting_started_progress', { cause: err }));
+    }
+}
+
+/**
+ * Gets getting started progress for the user, or creates it if it doesn't exist.
+ */
+export async function getOrCreateProgressByUser(user: DBUser, currentEnvironmentId: number): Promise<Result<GettingStartedProgressOutput>> {
+    return await db.knex.transaction(async (trx) => {
+        const existingResult = await getProgressByUserId(user.id, trx);
+
+        if (existingResult.isErr()) {
+            return Err(existingResult.error);
+        }
+
+        if (existingResult.value !== null) {
+            return Ok(existingResult.value);
+        }
+
+        const gettingStartedMeta = await getOrCreateMeta(user.account_id, currentEnvironmentId, trx);
+
+        if (gettingStartedMeta.isErr()) {
+            return Err(gettingStartedMeta.error);
+        }
+
+        const createdResult = await createProgress(
+            {
+                user_id: user.id,
+                getting_started_meta_id: gettingStartedMeta.value.id,
+                step: 0,
+                connection_id: null
+            },
+            trx
+        );
+
+        if (createdResult.isErr()) {
+            return Err(createdResult.error);
+        }
+
+        const newProgress = await getProgressByUserId(createdResult.value.user_id, trx);
+        if (newProgress.isErr()) {
+            return Err(newProgress.error);
+        }
+
+        if (newProgress.value === null) {
+            return Err('failed_to_get_getting_started_progress');
+        }
+
+        return Ok(newProgress.value);
+    });
+}
+
+/**
+ * Update getting started progress for a user and return the updated object.
+ */
+export async function patchProgressByUser(user: DBUser, input: PatchGettingStartedInput): Promise<Result<void>> {
+    return await db.knex.transaction(async (trx) => {
+        // Ensure meta/progress exist and fetch meta information (environment, integration)
+        const existing = await getProgressByUserId(user.id, trx);
+        if (existing.isErr()) {
+            return Err(existing.error);
+        }
+
+        if (existing.value === null) {
+            return Err('getting_started_progress_not_found');
+        }
+
+        const update: Partial<DBGettingStartedProgress> = {};
+
+        if (typeof input.step !== 'undefined') {
+            update.step = input.step;
+        }
+
+        if (typeof input.connection_id !== 'undefined') {
+            if (input.connection_id === null || input.connection_id === '') {
+                update.connection_id = null;
+            } else {
+                try {
+                    const { error, response: connection } = await connectionService.getConnection(
```

Comment:
that's why all functions that talk to the db should have a knex instance passed as argument. This query should be part of the transaction. I am not asking you to change it though.